### PR TITLE
profiling for physics simulation

### PR DESCRIPTION
**Pull request by AndrewMeadows**
This adds profiling to help debug performance problems in the physics engine.

### CODE REVIEW

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -975,6 +975,11 @@ void Application::keyPressEvent(QKeyEvent* event) {
                 _myAvatar->setDriveKeys(UP, 1.0f);
                 break;
 
+            case Qt::Key_F: {
+                _physicsEngine.dumpNextStats();
+                break;
+            }
+
             case Qt::Key_Asterisk:
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: Press F once to get a bug printout of the next simulation frame's profile.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2089a8c2dd12220544cfe0312a260af5e1bd3e15 available at: https://gist.github.com/c478ff36d3250ae4d5df<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4592.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4592.dmg

File: libraries/physics/src/PhysicsEngine.cpp [hunk 2]
```diff
@@ -287,66 +288,76 @@ void PhysicsEngine::init(EntityEditPacketSender* packetSender) {
...
-        // This is step (3) which is done outside of stepSimulation() so we can lock _entityTree.
-        //
-        // Unfortunately we have to unlock the simulation (above) before we try to lock the _entityTree
-        // to avoid deadlock -- the _entityTree may try to lock its EntitySimulation (from which this 
-        // PhysicsEngine derives) when updating/adding/deleting entities so we need to wait for our own
-        // lock on the tree before we re-lock ourselves.
-        //
-        // TODO: untangle these lock sequences.
-        _entityTree->lockForWrite();
+    {
```
Comment (sethalves) [inline libraries/physics/src/PhysicsEngine.cpp]: why the extra { } here?

File: libraries/physics/src/ThreadSafeDynamicsWorld.cpp [hunk 3]
```diff
@@ -26,6 +28,7 @@ ThreadSafeDynamicsWorld::ThreadSafeDynamicsWorld(
 }
 
 int	ThreadSafeDynamicsWorld::stepSimulation( btScalar timeStep, int maxSubSteps, btScalar fixedTimeStep) {
+    BT_PROFILE("stepSimulation");
```
Comment (sethalves) [inline libraries/physics/src/ThreadSafeDynamicsWorld.cpp]: indentation here seems strange

File: libraries/physics/src/PhysicsEngine.cpp (hunk 2 above)
Comment (AndrewMeadows) [inline libraries/physics/src/PhysicsEngine.cpp]: The extra braces are used to define the scope of the BT_PROFILE("stepSimulation") macro.  That macro creates a variable that starts a timer in its constructor and then it it destructor it measures the time elapsed and inserts that time into a static map of accumulators.  So the braces are used to force the variable to go out of scope when we're done doing the simulation work, but before we harvest all of the profiling info at the bottom when we call CProfileManager::dumpAll().

**Comment by hifi-gustavo** [comment]
Build console output for commit: 568050686db413f8dd6b2b8d7623e46b0a09fdcc available at: https://gist.github.com/c478ff36d3250ae4d5df<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4592.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4592.dmg

**Pull request merged**